### PR TITLE
Rework ImportModelClass to provide available headers

### PR DESCRIPTION
**Pull request by bayangan1991**
`ImporterModelForm` now provides a method that will present all the available mappable headers

calling `get_field_metadata` will return a dictionary of dataclasses that contain:

- `field` : The form field instance
- `help_text` : The `help_text` of the field, first pulled from `Meta.help_texts`, then the model fields
- `required`: Whether this field is required
- `sources`: A list of different headers that can be supplied to satisfy this field

This method is now used in place to generate the valid fields for the `FormClassBuilder`

### CODE REVIEW

File: djangomodelimport/fields.py [hunk 1]
```diff
@@ -22,12 +22,12 @@ class FlatRelatedField(forms.Field):
...
     """
 
-    def __init__(self, queryset, fields=[], *args, **kwargs):
+    def __init__(self, queryset, fields=None, *args, **kwargs):
         self.queryset = queryset
         # TODO: If lookup key is provided, allow using it to look up value instead of only
         # retrieving it off the object itself.
         self.model = queryset.model
-        self.fields = fields
+        self.fields = fields or {}
```
Comment (jarekwg) [inline djangomodelimport/fields.py:39]: Lols. It was defaulting to `list` when it should've been expecting `dict`..

File: djangomodelimport/formclassbuilder.py [hunk 2]
```diff
@@ -30,42 +31,45 @@ def valid_fields(self):
...
+            result: dict[str, list[list[str]]] = defaultdict(lambda: [[]])
+            for header in headers:
+                field_ = result[header.field_name]
+                field_[0].append(header.name)
+                for alt in header.alternatives:
+                    alt_fields = _flatten_headers(alt)
+                    for k, v in alt_fields.items():
+                        result[k].extend(v)
+
+            return dict(result)
```
Comment (jarekwg) [inline djangomodelimport/formclassbuilder.py]: This is a little hard to follow. Both the function's docstring and the code below it, with `field_` and `field_[0]`..

File: djangomodelimport/formclassbuilder.py [hunk 3]
```diff
@@ -30,42 +31,45 @@ def valid_fields(self):
...
+
+        # Get the viable headers for the importer class
+        form_headers = self.modelimportformclass.get_available_headers()
+
+        # Find all the valid field combinations
+        valid_headers = _flatten_headers(form_headers)
```
Comment (jarekwg) [inline djangomodelimport/formclassbuilder.py]: This is unclear as to what combinations it's talking about. I'm assuming it's referring to composite lookup widgets, but can we word it a bit better?

File: djangomodelimport/formclassbuilder.py [hunk 4]
```diff
@@ -30,42 +31,45 @@ def valid_fields(self):
...
+
+        # Create a header -> field lookup dictionary
+        field_lookup = {}
+        for field, header_group in valid_headers.items():
+            for headers in header_group:
+                field_lookup[frozenset(headers)] = field
+
+        # See if each valid field header if in the provided import headers
```
Comment (jarekwg) [inline djangomodelimport/formclassbuilder.py]: This one might need a reword too. Assuming you mean we're deciding which form fields are valid for use, based on the incoming headers?

File: djangomodelimport/formclassbuilder.py [hunk 5]
```diff
@@ -30,42 +31,45 @@ def valid_fields(self):
...
+        valid_present_fields = set()
+        for headers, field in field_lookup.items():
+            if headers & set(self.headers):
+                valid_present_fields.add(field)
+
+        # Add an extra JSON Fields as they are wily
```
Comment (jarekwg) [inline djangomodelimport/formclassbuilder.py]: :D

File: djangomodelimport/utils.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+@dataclasses.dataclass
+class ImportHeader:
+    field_name: str
+    field: Field
+    name: str
+    required: bool
+    display: str = ""
+    help_text: str = ""
+    alternatives: list[list["ImportHeader"]] = dataclasses.field(default_factory=list)
```
Comment (jarekwg) [inline djangomodelimport/utils.py]: Is it worth adding a couple of comments to these? Particularly `name` vs `field_name`, and `alternatives`.

File: djangomodelimport/forms.py [hunk 7]
```diff
@@ -42,3 +46,106 @@ def warnings(self):
...
+                    for new_sub_headers in map(
+                        partial(
+                            _get_headers,
+                            widget=None,
+                            field_instance=field_instance,
+                        ),
+                        sub_fields,
+                    ):
+                        found_headers.extend(new_sub_headers)
+                case HasSource(source=source):
```
Comment (jarekwg) [inline djangomodelimport/forms.py]: Very pretty!

**Review by jarekwg — COMMENTED** [review]
I'm being a little pedantic with comments here, just because this package is already a nightmare to return to each time we want to do something every 6 months or so.

Please review all comments and don't shy from making them overly beefy if that helps precisely explain what we're doing.

File: .gitignore [hunk 8]
```diff
@@ -14,3 +14,5 @@ environ/
 examples/example_site/env
 **.pem
 src/
+.idea/
+.vscode
```
Comment (bayangan1991) [inline .gitignore]: remove this

File: djangomodelimport/forms.py [hunk 9]
```diff
@@ -22,23 +26,126 @@ class ImporterModelForm(
...
+        """Generate a list of available fields for the ImporterClass"""
+        # 1) Evaluate Field type:
+        # - SourceFieldSwitcher: these are a collection of different ways to find a related object.
+        # - FlatRelatedField: these are a collection of other columns that build a relation on the fly.
+
+        # 2) For each field, check its widget to see what headers it reads from
+        # - Anything using NamedSourceWidget or has a "source" attr:
+        #     these lookup columns might not match the form field.
+        # - Anything using CompositeLookupWidget:
+        #     these lookup columns might not always mention the form field target.
```
Comment (bayangan1991) [inline djangomodelimport/forms.py:81]: remove

File: djangomodelimport/formclassbuilder.py [hunk 10]
```diff
@@ -1,74 +1,59 @@
...
         return self._get_modelimport_form_class(fields=form_fields)
 
     @cached_property
-    def valid_fields(self):
-        """Using the provided headers, prepare a list of valid fields for this importer.
-        Preserves field ordering as defined by the headers.
+    def valid_fields(self) -> list[str]:
+        """Using the available headers on the form, prepare a list of valid
+        fields for this importer. Preserves field ordering as defined by the headers.
         """
```
Comment (jarekwg) [inline djangomodelimport/formclassbuilder.py]: I think you might want to revert this docstring change -- the "provided headers" being described here are the ones coming in from the CSV import, not our header fetching code.

File: djangomodelimport/formclassbuilder.py [hunk 11]
```diff
@@ -1,74 +1,59 @@
...
+
+        # Get the viable headers for the importer class
+        form_field_metadata = self.modelimportformclass.get_field_metadata()
+
+        # Check each header combination against the input headers to
+        # see if they evaluate to a field
+        valid_present_fields = set()
+        for field_name, field_meta in form_field_metadata.items():
+            if isinstance(field_meta.field, (FlatRelatedField, JSONField)):
+                valid_present_fields.add(field_name)
```
Comment (jarekwg) [inline djangomodelimport/formclassbuilder.py:49]: A comment for why FlatRelatedField and JSONField get a free pass would help here. Can borrow old reasoning in this method.

File: djangomodelimport/forms.py [hunk 12]
```diff
@@ -22,23 +26,145 @@ class ImporterModelForm(
...
+        """
+        fields = cls.get_field_metadata()
+
+        header_set = {"id"}
+        headers = [("id", "ID")]
+
+        for field in fields.values():
+            for sources in field.sources:
+                for header, label in sources:
+                    if header not in header_set:
```
Comment (jarekwg) [inline djangomodelimport/forms.py:64]: Oh, so you _are_ deduping here already. Nice!

**Review by jarekwg — APPROVED** [review]
Love it! Great work!
Bit of a journey, but we got a pretty good level of untangling here, considering the complexity of this repo.

**Comment by nickmellor** [comment]
Hi Ryan,

I hope you're keeping well and Uptick is going strong also.

I still seem to be a member of this repo, about 6 years after leaving
Uptick.

Please consider removing me ;)

Regards,

Nick

Perhaps for security reasons you might want to remove me ;)

**Comment by jarekwg** [comment]
> Hi Ryan, I hope you're keeping well and Uptick is going strong also. I still seem to be a member of this repo, about 6 years after leaving Uptick. Please consider removing me ;) Regards, Nick Perhaps for security reasons you might want to remove me ;)

Hey @nickmellor -- this is a public repo. I've just reviewed membership, and you don't have write access.
I suspect you may just be subscribed to updates?

File: djangomodelimport/forms.py (hunk 12 above)
Comment (bayangan1991) [reply djangomodelimport/forms.py:64]: Yeah, de-duping the headers keys, but not the header labels.
